src/webgpu/capability_info.ts

```diff
@@ -295,56 +1243,124 @@ export const kCanvasColorSpacesInfo: {
 };
 export const kCanvasColorSpaces = keysOf(kCanvasColorSpacesInfo);
 
-/** Per-GPUTextureFormat info. */
-// Exists just for documentation. Otherwise could be inferred by `makeTable`.
-// MAINTENANCE_TODO: Refactor this to separate per-aspect data for multi-aspect formats. In particular:
-// - bytesPerBlock only makes sense on a per-aspect basis. But this table can't express that.
-//   So we put depth32float-stencil8 to be an unsized format for now.
-export type TextureFormatInfo = {
-  /** Whether the format can be used as `RENDER_ATTACHMENT`. */
-  renderable: boolean;
+/** Per-GPUTextureFormat-per-aspect info. */
+interface TextureFormatAspectInfo {
+  /** Whether the aspect can be used as `COPY_SRC`. */
+  copySrc: boolean;
+  /** Whether the aspect can be used as `COPY_DST`. */
+  copyDst: boolean;
+  /** Whether the aspect can be used as `STORAGE`. */
+  storage: boolean;
+  /** The "texel block copy footprint" of one texel block; `undefined` if the aspect is unsized. */
+  bytes: number | undefined;
+}
+/** Per GPUTextureFormat-per-aspect info for color aspects. */
+interface TextureFormatColorAspectInfo extends TextureFormatAspectInfo {
+  bytes: number;
+  /** "Best" sample type of the format. "float" also implies "unfilterable-float". */
+  type: 'float' | 'uint' | 'sint' | 'unfilterable-float';
+}
+/** Per GPUTextureFormat-per-aspect info for depth aspects. */
+interface TextureFormatDepthAspectInfo extends TextureFormatAspectInfo {
+  /** "depth" also implies "unfilterable-float". */
+  type: 'depth';
+}
+/** Per GPUTextureFormat-per-aspect info for stencil aspects. */
+interface TextureFormatStencilAspectInfo extends TextureFormatAspectInfo {
+  bytes: 1;
+  type: 'uint';
+}
+
+/**
+ * Per-GPUTextureFormat info.
+ * This is not actually the type of values in kTextureFormatInfo; that type is fully const
+ * so that it can be narrowed very precisely at usage sites by the compiler.
+ * This type exists only a type check on the inferred type of kTextureFormatInfo.
+ * Documentation is also written here, but not actually visible to the IDE.
```

Comment:
Ah... this is unfortunate... Is this fixable in follow up CLs once we aren't in between states?

Reply:
Unlikely. I didn't realize this before this PR, but we actually already had this problem.

We need the type of the entire object to be const because we rely on having the exact type of every format for a bunch of type narrowing that happens elsewhere - like if you call `kTextureFormat[format]` where `format` is of type `"rgba8unorm" | "bgra8unorm"` then a bunch of the fields that are normally optional become guaranteed.

But then because every single piece of data in the entire structure has its own place in the type, every single one of those would need the documentation attached to it individually. Which would require defining the entire type explicitly (redundantly) and duplicating the documentation all over.

I think typedoc would need a pretty substantial piece of magic to let us declare the documentation in a way that will distribute it throughout the inferred type for us, https://github.com/microsoft/tsdoc/issues/241 probably.

Reply:
for the record, this turned out to be solvable: #3351